### DISCUSSION TITLE
Is it safe to increment number on row creation?

### DISCUSSION THREAD

**Discussion by DominikGrodl** [supabase/supabase#15709] [Questions] [answered]
Hi, I have a question which may have a simple answer. I am not a backend engineer, so sorry if the question seems dumb 😄 

Is it safe to increment a number in a row when a row in different table is created? Imagine this situation (not my context, but the easiest to imagine):

I have a table of posts, where every record has a column named "number_of_likes". When another user likes the post, a record in the "likes" table is created and this triggers a function which increments the number_of_likes on a given post record. The same but with decrement happens when a like is removed. Is this safe to do? What would happen if many "likes" record get created very fast, is it possible that some of them would "fall through" and not increment? And what if many likes and removals come, is it possible to have a race condition where one of the triggers tries to update stale data, or are the requests to increment/decrement handled in a way which would prevent that?

Is there possibly a different way to achieve the same goal without these problems? I know we can get the count of records with given predicate, but I would rather find a more streamlined way if there is one.

Thanks for any help and sorry again if this is a dumb question.

**Comment by GaryAustin1** [comment, marked as answer]
Using a trigger function to do this is reasonable.  You should not lose anything because if the trigger function fails, the insert/update to the main table will also fail.   

You probably will need to make these functions 'security definer' type as I assume you have RLS to keep the user from changing the likes with their own call.   Then you might want to adjust the grants on the function so users can't call them from the API.

https://supabase.com/docs/guides/database/functions#security-definer-vs-invoker has some basic info on security definer functions.

**Reply by DominikGrodl** [reply]
Awesome, thanks for the info. I have them defined as security definer already, so this should be fine. Thanks again!

**Reply by GaryAustin1** [reply]
Oh, if they are trigger functions then you don’t want to do the grants as the can’t be called from the API directly.

**Reply by DominikGrodl** [reply]
Oh okay, I will take a look at that, honestly all the security details are little over my head 😄 I will take a look at the link you sent, thanks for that 🙂
